server/server.go

```diff
@@ -6,53 +6,299 @@ package server
+				PersistedIndex: *change.persistedIndex,
+			})
+			if err != nil {
+				return fmt.Errorf("failed to send the update: %w", err)
 			}
-			break
+
+		case <-producer.Context().Done():
+			return fmt.Errorf("producer context: %w", producer.Context().Err())
+
+		case <-serv.polling.ctx.Done():
+			return fmt.Errorf("server context: %w", serv.polling.ctx.Err())
 		}
-		res := messages.EventResult{EventId: evt.EventId, Timestamp: pbts.Now()}
-		results = append(results, &res)
 	}
-	return &messages.PublishReply{Results: results}, nil
 }
 
-// StreamAcknowledgements is the server implementation of the gRPC StreamAcknowledgements call
-func (serv shipperServer) StreamAcknowledgements(streamReq *messages.StreamAcksRequest, prd pb.Producer_StreamAcknowledgementsServer) error {
+// Close implements the Closer interface
+func (serv *shipperServer) Close() error {
+	serv.close.Do(func() {
+		// polling must be stopped first, otherwise it would try to write
+		// a notification to a closed channel and this would cause a panic
+		serv.polling.stop()
+		<-serv.polling.stopped
+
+		serv.logger.Debug("shutting down all notifications...")
+		serv.notifications.shutdown()
+		serv.logger.Debug("all notifications have been shut down")
+	})
+
+	return nil
+}
+
+func (serv *shipperServer) startPolling() {
+	go func() {
+		ticker := time.NewTicker(serv.cfg.PollingInterval)
+		defer ticker.Stop()
 
-	// we have no outputs now, so just send a single dummy event
-	evt := messages.StreamAcksReply{Acks: []*messages.Acknowledgement{{Timestamp: pbts.Now(), EventId: streamReq.Source.GetInputId()}}}
-	err := prd.Send(&evt)
+		for {
+			select {
 
+			case <-serv.polling.ctx.Done():
+				err := serv.polling.ctx.Err()
+				if err != nil && errors.Is(err, context.Canceled) {
+					serv.logger.Error(err)
+				}
+				close(serv.polling.stopped) // signaling back to `Close`
+				return
+
+			case <-ticker.C:
+				serv.logger.Debug("updating indices...")
+				err := serv.updateIndices(serv.polling.ctx)
+				if err != nil {
+					serv.logger.Errorf("failed to update indices: %s", err)
+				} else {
+					serv.logger.Debug("successfully updated indices.")
+				}
+			}
+		}
+	}()
+}
+
+// updateIndices updates in-memory indices and notifies subscribers if necessary.
+// TODO: this is a temporary implementation until the queue supports the `persisted_index`.
+func (serv *shipperServer) updateIndices(ctx context.Context) error {
+
+	// TODO: for now we calculate an approximate value
+	// we cannot rely on this value and needs to be changed when the queue is ready.
+	metrics, err := serv.queue.Metrics()
 	if err != nil {
-		return fmt.Errorf("error in StreamAcknowledgements: %w", err)
+		return fmt.Errorf("failed to fetch queue metrics: %w", err)
 	}
+
+	log := serv.logger
+	c := change{}
+
+	if metrics.UnackedConsumedEvents.Exists() {
+		oldPersistedIndex := serv.GetPersistedIndex()
+		persistedIndex := serv.GetAcceptedIndex() - int64(metrics.UnackedConsumedEvents.ValueOr(0))
```

Comment:
Maybe this is leftover, but like we talked about on the issue, `UnackedConsumedEvents` has no connection to the persisted index, if we want the delta here then we need to save the previous value between iterations.

Reply:
It is a left-over. I initially opened this draft PR to demonstrate the current state for our discussion.